op-program/host/config/config.go

```diff
@@ -56,7 +56,13 @@ type Config struct {
 	L2Head common.Hash
 	// L2OutputRoot is the agreed L2 output root to start derivation from
 	L2OutputRoot common.Hash
-	L2URL        string
+	// L2URL is the URL of the L2 node to fetch L2 data from, this is the canonical URL for L2 data
+	// in the case of L2ExperimentalEnabled = true, this URL is used as a fallback if the experimental URL fails or cannot retrieve the desired data
+	L2URL string
+	// L2ExperimentalURL is the URL of the L2 node (non hash db archival node, for example, reth archival node) to fetch L2 data from
+	L2ExperimentalURL string
```

Comment:
I'd just use a single `L2URL`. The fallback is kind of cool but I think we're just adding unnecessary complexity. The `L2ExperimentalEnabled` flag is useful so we can disable using witnesses entirely.

Then we can first enable witnesses on our vm-runner box that just runs cannon/op-program repeatedly outside of dispute games first. That can be pointed at an experimental version of op-geth if needed and it will build up confidence that it all works properly.  Then we can actually enable it for the real op-challenger.

Reply:
Kept this in for now, but should be easy to remove if we decide to remove it.

Reply:
Sorry I wrote an answer to this yesterday and somehow managed to not post it.

`vm-runner` is actually the `op-challenger run-trace` subcommand which can run the trace provider op-challenger uses over and over based on real chain data. We have a couple of boxes in the cloud that run it pointed at op-sepolia and op-mainnet to verify various different configurations correctly validate the chain. So we run cannon with the op-program prestate actually specified in the `FaultDisputeGame` on chain, cannon with an updated op-program prestate similar to what will be deployed with holocene (though holocene not yet active), a multithreaded cannon prestate and asterisc/kona.

There's two problems with trying to deploy this to production networks:
1. You need to update the prestate used on chain so it will emit the new hints. That requires a governance vote to approved the changes.
2. cannon never actually runs in production because dispute games only get down to the split depth when we deliberately force it to (and spend a lot of ETH in the process).

So I definitely like the idea of a feature toggle to control whether or not we call executionWitness but I don't actually see the need for a separate L2 URL.

Reply:
I see, our case is a little bit different as we want to use reth as the experimental back and geth as the fallback due to geth state growth issues we're facing. Basically our disk size is projected to exceed 30T in Q1 next year and after that point it's gonna be super hard to run a geth hashdb archival node. 

That's why we need a different url to provide experimental apis

Reply:
> I see, our case is a little bit different as we want to use reth as the experimental back and geth as the fallback

Ah yeah, that a good reason to justify the extra code. Makes sense.

> From Dockerfile.repro, it seems like any commits to op-program could potentially change the reproducible-prestate? Curious what actually changes the absolute prestate?

Yes, any changes to the op-program client will result in a different prestate generated from it. Typically I assume that to get the exact same prestate you have to be on the same commit.  But to actually use a different prestate on chain, that new version needs to be set as the absolute prestate in the `FaultDisputeGame` contract on L1. That requires the `ProxyAdminOwner` to sign a transaction and should only be done for governance approved versions.

> This is expected as it's very expensive, is there a good way to test this feature change? My understanding is that cannon execution into specific steps does not change anything as the block we execute would get all the requried input from execution witness and eth_getProof and have them cached in the oracle, no?

Because there's a new hint type emitted from the client, this will require a new prestate so will change the specific steps that cannon executes. We can switch back and forth between how we fetch the data without any issues once that extra hint is present, it's just adding the hint that changes the actual execution. We'll need to update the prestate with Holocene to support that hard fork anyway so it's pretty likely that emitting the new hint can ship with that (and then we can decide when to depend on executionWitness to actually fetch the data independently).

So the way I'd test this is with the `run-trace` subcommand. We can easily deploy a new prestate there that emits the hint and still have it verify the actual L2 blocks. It won't produce the same final cannon state as the on-chain version but it should still correctly determine if the proposed output root is valid or not. We can also monitor the logs for any cases where it needs to fall back to `debug_dbGet`.